test/pkg/compose/templatingdata/assembler.go

```diff
@@ -0,0 +1,392 @@
+package templatingdata
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"net/http"
+	"os"
+	"path"
+	"path/filepath"
+	"regexp"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"github.com/openshift/microshift/pkg/util"
+	"github.com/openshift/microshift/test/pkg/testutil"
+	"k8s.io/klog/v2"
+)
+
+var (
+	releaseInfoRpmRx   = regexp.MustCompile("^microshift-release-info-.*.rpm$")
+	errNoRemoteRelease = fmt.Errorf("release from remote repo not found")
+)
+
+type TemplatingDataOpts struct {
+	ArtifactsMainDir               string
+	TemplatingDataFragmentFilepath string
+	SkipContainerImagesExtraction  bool
+}
+
+func (o *TemplatingDataOpts) Construct() (*TemplatingData, error) {
+	klog.InfoS("Constructing TemplatingData")
+
+	rpmRepos := path.Join(o.ArtifactsMainDir, "rpm-repos")
+	localRepo := path.Join(rpmRepos, "microshift-local")
+	fakeNextRepo := path.Join(rpmRepos, "microshift-fake-next-minor")
+	baseRepo := path.Join(rpmRepos, "microshift-base")
+	externalRepo := path.Join(rpmRepos, "microshift-external")
+
+	if exists, err := util.PathExists(localRepo); err != nil {
+		return nil, fmt.Errorf("failed to check if %s exists: %w", localRepo, err)
+	} else if !exists {
+		return nil, fmt.Errorf("%s does not exist - did you run build_rpms.sh and create_local_repos.sh?", localRepo)
+	}
+
+	var td *TemplatingData
+	var err error
+
+	if o.TemplatingDataFragmentFilepath != "" {
+		td, err = unmarshalTemplatingData(o.TemplatingDataFragmentFilepath)
+		if err != nil {
+			return nil, err
+		}
+		klog.InfoS("TemplatingData fragment was provided and loaded", "intermediateTeplatingData", td)
+	} else {
+		td = &TemplatingData{}
+	}
+
+	td.Arch = getArch()
+
+	if td.Source.Repository == "" {
+		td.Source, err = o.getReleaseFromLocalFs(localRepo)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if td.Base.Repository == "" {
+		td.Base, err = o.getReleaseFromLocalFs(baseRepo)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if td.FakeNext.Repository == "" {
+		td.FakeNext, err = o.getReleaseFromLocalFs(fakeNextRepo)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if td.External.Repository == "" {
+		exists, err := util.PathExistsAndIsNotEmpty(externalRepo)
+		if err != nil {
+			return nil, err
+		}
+		if exists {
+			td.External, err = o.getReleaseFromLocalFs(externalRepo)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	if td.Current.Repository == "" {
+		td.Current, err = o.getReleaseFromRemoteRepo(td.Source.Minor)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if td.Previous.Repository == "" {
+		td.Previous, err = o.getReleaseFromRemoteRepo(td.Source.Minor - 1)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if td.YMinus2.Repository == "" {
+		td.YMinus2, err = o.getReleaseFromRemoteRepo(td.Source.Minor - 2)
```

Comment:
If we set a precondition that the repository needed to be configured in the OS, would we need separate functions for getting the version from a local or remote repo?

Reply:
I was thinking about how the tool configures image builder with repos it can use. It would be possible to similarly template repo configurations for the OS. Maybe it's not worth the trouble to standardize, though. Or maybe there's room for another abstraction within the code for an RPMRepository interface with local and remote implementations.